Blish HUD/Controls/MultilineTextBox.cs

```diff
@@ -0,0 +1,237 @@
+﻿using System;
+using Blish_HUD.Input;
+using Microsoft.Xna.Framework;
+using Microsoft.Xna.Framework.Graphics;
+
+namespace Blish_HUD.Controls {
+    public class MultilineTextBox : TextInputBase {
+
+        private const int TEXT_TOPPADDING  = 7;
+        private const int TEXT_LEFTPADDING = 10;
+
+        public MultilineTextBox() {
+            _multiline = true;
+            _maxLength = 524288;
+        }
+
+        protected override void MoveLine(int delta) {
+            int newIndex = 0; // if targetLine is < 0, we set cursor index to 0
+
+            string[] lines = _text.Split(NEWLINE);
+
+            var cursor = GetSplitIndex(_cursorIndex);
+
+            int targetLine = cursor.Line + delta;
+
+            if (targetLine >= lines.Length) {
+                newIndex = _text.Length;
+            } else if (targetLine >= 0) {
+                float cursorLeft   = MeasureStringWidth(lines[cursor.Line].Substring(0, cursor.Character));
+                float minOffset    = cursorLeft;
+                int   currentIndex = 0;
+
+                var glyphs = _font.GetGlyphs(lines[targetLine]);
+
+                foreach (var glyph in glyphs) {
+                    float localOffset = Math.Abs(glyph.Position.X - cursorLeft);
+
+                    if (localOffset < minOffset) {
+                        newIndex  = currentIndex;
+                        minOffset = localOffset;
+                    }
+
+                    currentIndex++;
+                }
+
+                for (int i = 0; i < targetLine; i++) {
+                    newIndex += lines[i].Length + 1;
+                }
+            }
+
+            UserSetCursorIndex(newIndex);
+            UpdateSelectionIfShiftDown();
+        }
+
+        protected override void OnClick(MouseEventArgs e) {
+            base.OnClick(e);
+
+            int newIndex = GetCursorIndexFromFromPosition(this.RelativeMousePosition - new Point(TEXT_LEFTPADDING, TEXT_TOPPADDING));
+
+            if (_cursorIndex == newIndex && e.IsDoubleClick) {
+                SelectAll();
```

Comment:
Shouldn't this select the word instead of the hole text?

Reply:
Copied from above:

The answer to this is _yes_, but I don't think it will be included in this PR.  Should just be a matter of searching in both directions for the word boundary and updating the selection and would probably be a method on `TextInputBase` since both would use it so I'll make note to have it added in the future.

Reply:
Just kidding.  I went ahead and got it implemented.  🙂